docs: Add React 18 Upgrade to v8 upgrade guide (#1857)
Workday/canvas-kit: modules/docs/mdx/8.0-UPGRADE-GUIDE.mdx

### React 18 Upgrade

We've upgraded Canvas Kit to [React 18](https://reactjs.org/blog/2022/03/29/react-v18.html). This
change will not impact users who are already on v7 and are looking to upgrade to v8 as v8 will
support the same React versions as v7 (React 16.14 and 17.X) while still giving them the ability to

> upgrade to React 18 as needed. Although we didn't use any new features from React 18 in v8, _future_

Reviewer: I love the "_future_" note here!

### React 18 Upgrade

> In this release we upgraded React to version 18. [React 18](https://reactjs.org/blog/2022/03/29/react-v18.html) introduced several new features, but the
> most impactful change is the new [concurrent render API](https://reactjs.org/blog/2022/03/29/react-v18.html#what-is-concurrent-react). In the new API, you'll use the `.createRoot()`
> method which replaces `.render()`, which adds all of the improvements of React 18 and allows you to
> use concurrent features. Without it React behaves the same as it did in version 17.

Reviewer 2: Wordsmithing (for consistency and to remove unnecessary text). 
```suggestion
We've upgraded Canvas Kit to [React 18](https://reactjs.org/blog/2022/03/29/react-v18.html). React
18 introduces several new features, but the most impactful change is the new
[concurrent render API](https://reactjs.org/blog/2022/03/29/react-v18.html#what-is-concurrent-react).
The new `createRoot` method replaces `render` and gives you access to the improvements in React 18
including the concurrent features. Without it, React behaves as it did in version 17.
```

Canvas Kit doesn’t use the render API directly in our packages but Storybook uses it to render
Canvas Kit components. Because of this, we upgraded Storybook dependencies to support React 18.

> **Impact on CK:** Updating to React 18 didn't cause any breaking changes in our code. However, we did have to update some types based on some type packages that where upgraded.

Reviewer 3: I wouldn't use an acronym here.

```suggestion
**Impact on Canvas Kit:** Updating to React 18 didn't cause any breaking changes in our code. However, we did have to update some types based on some type packages that where upgraded.
```

> To see React v18 features in action we had to upgrade Storybook to v6.5.x.

Reviewer 2: Thoughts on removing the **Impact on CK** section? I'm not sure we need it in the Upgrade Guide. If we feel it's important, we could include a link to the [GH discussion](https://github.com/Workday/canvas-kit/discussions/1545).

**Impact on users:** Canvas Kit on React 18 doesn't produce any problem if it runs on < 18 version
react environment. However, users will see a  console warning about `.root()` being deprecated. It's possible for
users to use Canvas Kit v8 without upgrading to React 18 but some features may be ignored and limited

> by the legacy root API usage.

Reviewer 2: I liked how the official announcement was clear about the React 18 upgrade not impacting users.

> Canvas Kit will continue to support React 16.14, React 17.X, and React 18.X  in v8.
>
> This change will not impact product teams that are already on v7 and are looking to upgrade to v8 as v8 will support the same React versions as v7 while still giving teams the ability to upgrade to React 18 as needed.

What you do think about reusing that language here and integrating it with what you've already written regarding the console warning and the limitations of using Canvas Kit v8 without React 18?

Also, can we be more specific about which features "may be ignored and limited by the legacy root API usage"? I assume Canvas Kit will work just the same for users who are still on React 16.14 or 17 -- are we referring to features which may be added to Canvas Kit in the future?

### React 18 Upgrade

> In this release we upgraded React to version 18. React 18 introduced several new features, but the

Reviewer 4: ```suggestion
In this release we upgraded React to version 18. [React 18](https://reactjs.org/blog/2022/03/29/react-v18.html) introduced several new features, but the
```

> most impactful change is the new concurrent render API. In the new API, you'll use the .createRoot()

Reviewer 4: ```suggestion
most impactful change is the new [concurrent render API](https://reactjs.org/blog/2022/03/29/react-v18.html#what-is-concurrent-react). In the new API, you'll use the .createRoot()
```

> most impactful change is the new [concurrent render API](https://reactjs.org/blog/2022/03/29/react-v18.html#what-is-concurrent-react). In the new API, you'll use the .createRoot()

Reviewer 4: ```suggestion
most impactful change is the new [concurrent render API](https://reactjs.org/blog/2022/03/29/react-v18.html#what-is-concurrent-react). In the new API, you'll use the `.createRoot()`
```

> method which replaces .render(), which adds all of the improvements of React 18 and allows you to

Reviewer 4: ```suggestion
method which replaces `.render()`, which adds all of the improvements of React 18 and allows you to
```

use concurrent features. Without it React behaves the same as it did in version 17.

> Canvas Kit doesn’t use the render API directly in our packages. But Storybook uses it to render

Reviewer 4: ```suggestion
Canvas Kit doesn’t use the render API directly in our packages but Storybook uses it to render
```

> Canvas Kit components. So we need to have Storybook dependencies supporting React 18 as well.

Reviewer 4: ```suggestion
Canvas Kit components. Because of this, we upgraded Storybook dependencies to support React 18.
```

> **Impact on CK:** Updating to React 18 didn't cause any breaking changes in our code (only switching

Reviewer 4: ```suggestion
**Impact on CK:** Updating to React 18 didn't cause any breaking changes in our code. However, we did have to update some types based on some type packages that where upgraded.
```

> to new types). To see v18 features in action it required updating Storybook to v6.5.x supporting

Reviewer 4: ```suggestion
To see React v18 features in action we had to upgrade Storybook to v6.5.x.```

> React 18.

Reviewer 4: ```suggestion
```

**Impact on users:** Canvas Kit on React 18 doesn't produce any problem if it runs on < 18 version

> react environment. Users will got console warning about deprecated legacy root. So it's possible for

Reviewer 4: ```suggestion
react environment. However, users will see a  console warning about `.root()` being deprecated. It's possible for
```

> users to use Canvas Kit 8 without upgrade to React 18 but some features can be ignored and limited

Reviewer 4: ```suggestion
users to use Canvas Kit v8 without upgrading to React 18 but some features may be ignored and limited
```

### React 18 Upgrade

We've upgraded Canvas Kit to [React 18](https://reactjs.org/blog/2022/03/29/react-v18.html). React
18 introduces several new features, but the most impactful change is the new
[concurrent render API](https://reactjs.org/blog/2022/03/29/react-v18.html#what-is-concurrent-react).
The new `createRoot` method replaces `render` and gives you access to the improvements in React 18
including the concurrent features. Without it, React behaves as it did in version 17.

Canvas Kit doesn’t use the render API directly in our packages but Storybook uses it to render
Canvas Kit components. Because of this, we upgraded Storybook dependencies to support React 18.

> This change will not impact product teams that are already on v7 and are looking to upgrade to v8 as

Reviewer: This is looking really good especially with the latest change. I recommend moving this section up to be directly after the first sentence that starts and ends on line 81. 

The content after that sentence would start a new paragraph.

I recommend this to make sure that there is little or no room for confusion around impact to users. Users will be looking for what this means to them as soon as they hear that we're upgraded to React 18.

Reviewer 2: Amending the above suggestion based on conversation with @RayRedGoose:

```
### React 18 Upgrade

We've upgraded Canvas Kit to [React 18](https://reactjs.org/blog/2022/03/29/react-v18.html). This
change will not impact users who are already on v7 and are looking to upgrade to v8 as v8 will
support the same React versions as v7 (React 16.14 and 17.X) while still giving them the ability to
upgrade to React 18 as needed. Although we didn't use any new features from React 18 in v8, _future_
updates to Canvas Kit may require upgrading to React 18 to support them.

React 18 introduces several new features...
```

v8 will support the same React versions as v7 while still giving teams the ability to upgrade to
React 18 as needed. Canvas Kit will continue to support React 16.14, React 17.X in v8, so it allows

> to upgrade to Canvas Kit v8 without upgrading to React 18. But some new features may be ignored and

Reviewer: > But some new features may be ignored and
limited by the legacy root API usage and a console warning about `.root()` being deprecated will be
presented.

I'm also wondering what features could be ignored and I couldn't find more info on a `.root()` console warning. My understanding is that if users are on React 16.14+ or 17, they will build Canvas Kit successfully without any console warnings or ignored features.

limited by the legacy root API usage and a console warning about `.root()` being deprecated will be
presented.

> You can read more about this upgrade in the
> [discussion](https://github.com/Workday/canvas-kit/discussions/1545)

Reviewer 2: Updating for consistency with language used in previous Upgrade Guides.
```suggestion
See our [React 18 discussion](https://github.com/Workday/canvas-kit/discussions/1545) for more
information.
```
